daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -811,35 +810,39 @@ class DirectOrBufferedDataOutputStream private[io] (
 
   protected def deliverContent(directDOS: DirectOrBufferedDataOutputStream, finfo: FormatInfo) = {
 
-    val ba = this.bufferingJOS.getBuf
+    val bufOS = this.bufferingJOS
     val bufferNBits = this.relBitPos0b // don't have to subtract a starting offset. It's always zero in buffered case.
 
     if (directDOS.isEndOnByteBoundary && this.isEndOnByteBoundary) {
       // no fragment bytes anywhere - just take the bytes
 
       val nBytes = (bufferNBits / 8).toInt
       val nBytesPut = {
-        if (this.bufferingJOS.isFile)
-          directDOS.putFile(
-            this.bufferingJOS.tempFile.toPath,
-            this.bufferingJOS.bytesWritten * 8, 4096, finfo)
-        else
-          directDOS.putBytes(ba, 0, nBytes, finfo)
+        if (bufOS.isFile) {
+          bufOS.flush
+          val nBitsPut = directDOS.putFile(
+            bufOS.tempFile.toPath,
+            bufOS.nBytes * 8, chunkSizeInBytes, finfo)
```

Comment:
I'm not sure this is correct for blobs. Although blob files are multiples of bytes, we actually support fragment bytes from the blob file. So I think instead of nBytes it should be bufferNBits. The assert probably also wants to be different since putBytes returns the number of bytes put, but putFile returns the number of bits, which isn't always a multiple of 8.